runtime/src/main/java/org/corfudb/runtime/view/stream/StreamAddressSpace.java

```diff
@@ -0,0 +1,288 @@
+package org.corfudb.runtime.view.stream;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Function;
+
+import lombok.Synchronized;
+
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.StreamOptions;
+
+/**
+ * This is an abstract class which represents the space of addresses of a stream. It is based on the
+ * assumption that the address space of a stream can be traversed forward and backwards.
+ *
+ * The abstract method 'findAddresses' defines how to traverse the space of unexplored addresses,
+ * i.e, addresses between the latest synced point (oldTail) and the latest update (newTail).
+ * For example, the implementation of BackpointerAddressSpace assumes the use of backpointers
+ * to determine the previous log entry for a stream.
+ *
+ * Created by amartinezman on 4/23/18.
+ */
+public abstract class StreamAddressSpace implements IStreamAddressSpace {
+
+    protected CopyOnWriteArrayList<Long> addresses;
+    protected UUID streamId;
+    protected CorfuRuntime runtime;
+    protected int maxInd;
+    protected int currInd;
+    protected StreamOptions options;
+
+    public StreamAddressSpace(UUID id, CorfuRuntime runtime) {
+        this.streamId = id;
+        this.runtime = runtime;
+        this.addresses = new CopyOnWriteArrayList();
+        this.currInd = -1;
+        this.maxInd = -1;
+    }
+
+    @Override
+    public void setStreamOptions(StreamOptions options) {
+        this.options = options;
+    }
+
+    @Override
+    public void reset() {
+        currInd = -1;
+    }
+
+    @Override
+    public void seek(long address) {
+
+        int tmpPtr = addresses.indexOf(address);
+
+        if (tmpPtr != -1) {
+            currInd = tmpPtr;
+            return;
+        }
+
+        throw new RuntimeException("Couldn't seek " + streamId + " to address " + address);
```

Comment:
probably need to make this more specific, maybe call it NonSeekableAddressException ?